src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -88,271 +107,399 @@ private void PublishTelemetry(string message, string taskId)
+                }
+            }
 
-                    if (usedCpuPercentage >= CPU_UTILIZATION_PERCENTAGE_THRESHOLD)
+            if (PlatformUtil.RunningOnLinux)
+            {
+                // Some compact Linux distributions like UBI may not have "free" utility installed, or it may have a custom output
+                // We don't want to break currently existing pipelines with ADO warnings
+                // so related errors thrown here will be sent to the trace or debug logs by caller methods
+
+                using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+                {
+                    processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
                     {
-                        string message = $"CPU utilization is higher than {CPU_UTILIZATION_PERCENTAGE_THRESHOLD}%; currently used: {usedCpuPercentage:0.00}%";
+                        if (!message.Data.StartsWith("Mem:"))
+                        {
+                            return;
+                        }
 
-                        PublishTelemetry(message, taskId);
+                        var processInvokerOutputString = message.Data;
+                        var memoryInfoString = processInvokerOutputString.Split(" ", StringSplitOptions.RemoveEmptyEntries);
 
-                        break;
-                    }
+                        if (memoryInfoString.Length != 7)
+                        {
+                            throw new Exception("\"free\" utility has non-default output");
+                        }
 
+                        lock (_memoryInfoLock)
+                        {
+                            _memoryInfo.Updated = DateTime.Now;
+                            _memoryInfo.TotalMemoryMB = Int32.Parse(memoryInfoString[1]);
+                            _memoryInfo.UsedMemoryMB = Int32.Parse(memoryInfoString[2]);
+                        }
+                    };
+
+                    processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
+                    {
+                        Trace.Error(message.Data);
+                    };
+
+                    try
+                    {
+                        var filePath = "free";
+                        var arguments = "-m";
+                        await processInvoker.ExecuteAsync(
+                                workingDirectory: string.Empty,
+                                fileName: filePath,
+                                arguments: arguments,
+                                environment: null,
+                                requireExitCodeZero: true,
+                                outputEncoding: null,
+                                killProcessOnCancel: true,
+                                cancellationToken: _context.CancellationToken);
+                    }
+                    catch (Win32Exception ex)
```

Comment:
In case if debug flag is set, it will be included into metrics string, e.g 
#[debug]Agent environment resources - Disk: / Available 878694.00 MB out of 1031018.00 MB, Unable to get memory info, exception: "free" utility is unavailable. Exception: An error occurred trying to start process 'free -m' with working directory '/home/runner/azure-pipelines-agent/_layout/linux-x64/bin'. No such file or directory, ...